sdk/freestanding-execution-engine/src/main.rs

```diff
@@ -337,103 +317,109 @@ fn main() -> Result<(), Box<dyn Error>> {
-            StandardStream::Stderr(file_rights) => (file_rights.file_name(), file_rights.rights()),
-        };
-        let rights = Rights::try_from(*rights as u64)
-            .map_err(|e| format!("Failed to convert u64 to Rights: {:?}", e))?;
-        file_table.insert(PathBuf::from(path), rights);
+    // Add write permission to output path
+    for file_path in cmdline.output_sources.iter() {
+        // NOTE: inject the root path.
+        file_table.insert(Path::new("/").join(file_path), write_right);
     }
-    file_table.insert(PathBuf::from(OUTPUT_FILE), write_right);
-    right_table.insert(Principal::Program(prog_file_name.to_string()), file_table);
+    let program_id = Principal::Program(
+        prog_file_abs_path
+            .to_str()
+            .ok_or("Failed to convert program path to a string.")?
+            .to_string(),
+    );
+    right_table.insert(program_id.clone(), file_table);
+    info!("The final right tables: {:?}", right_table);
 
-    let vfs = Arc::new(Mutex::new(FileSystem::new(right_table, &std_streams_table)));
-    vfs.lock()
-        .map_err(|e| format!("Failed to lock vfs, error: {:?}", e))?
-        .write_file_by_filename(
-            &Principal::InternalSuperUser,
-            &prog_file_name,
-            &program,
-            false,
-        )?;
-    info!("WASM program {} loaded into VFS.", prog_file_name);
+    let mut vfs = FileSystem::new(right_table)?;
 
-    load_data_sources(&cmdline, vfs.clone())?;
+    load_input_sources(&cmdline.input_sources, &mut vfs)?;
     info!("Data sources loaded.");
 
-    info!("Invoking main.");
+    info!("Invoking main on {:?}.", prog_file_abs_path);
     let main_time = Instant::now();
     let options = Options {
         environment_variables: cmdline.environment_variables,
         program_arguments: cmdline.program_arguments,
         enable_clock: cmdline.enable_clock,
         ..Default::default()
     };
+    let program = vfs.read_file_by_absolute_path(prog_file_abs_path)?;
     let return_code = execute(
         &cmdline.execution_strategy,
-        vfs.clone(),
-        &prog_file_name,
+        vfs.spawn(&program_id)?,
+        program,
         options,
     )?;
     info!("return code: {:?}", return_code);
     info!("time: {} micro seconds", main_time.elapsed().as_micros());
 
     // Dump contents of stdout
     if cmdline.dump_stdout {
-        let buf = vfs
-            .lock()
-            .map_err(|e| format!("Failed to lock vfs, error: {:?}", e))?
-            .read_file_by_filename(&Principal::InternalSuperUser, "stdout")?;
-        let stdout_dump = std::str::from_utf8(&buf)
-            .map_err(|e| format!("Failed to convert byte stream to UTF-8 string: {:?}", e))?;
-        print!("{}", stdout_dump);
+        let buf = vfs.read_stdout()?;
+        let stdout_dump = std::str::from_utf8(&buf)?;
+        print!(
+            "---- stdout dump ----\n{}---- stdout dump end ----\n",
+            stdout_dump
+        );
     }
 
```

Comment:
I think you might need `io::stdout()::flush()` or something like that to avoid the two getting mixed up if `stdout` and `stderr` are going to the same place which isn't a terminal.

Whenever you have finished writing to `stdout` and are about to write to `stderr` (and perhaps the other way round if you are feeling nervous) would probably be a good time to `flush`.